File: src/java/dp/B_12852.java
package java.dp;

import java.io.*;

public class B_12852 {
	static int[] dp;
	static int[] str;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		
		dp = new int[n + 1];
		str = new int[n + 1];
		
		System.out.println(getMinOper(n));
		
		StringBuilder sb = new StringBuilder();
		
		while(n != 0) {
			sb.append(n).append(' ');
			n = str[n];
		}
		
		System.out.println(sb);
	}
	
	static int getMinOper(int n) {
		if (dp[n] != 0 || n == 1) {
			return dp[n];
		}
		
		str[n] = n - 1;
		int temp = getMinOper(n - 1);
		
		if (n % 3 == 0) {
			int result = getMinOper(n / 3);
			if (temp > result) {
				temp = result;
				str[n] = n / 3;
			}
		}
		if (n % 2 == 0) {
			int result = getMinOper(n / 2);
			if (temp > result) {
				temp = result;
				str[n] = n / 2;
			}
		}

		return dp[n] = temp + 1;
	}

}

File: src/kotlin/dp/B_12852.kt
package kotlin.dp

class B_12852 {
    private val max = 1_000_001
    private lateinit var dp: IntArray
    private lateinit var next: IntArray

    operator fun invoke() {
        val n = System.`in`.bufferedReader().readLine().toInt()
        dp = IntArray(n + 1) { max }
        dp[1] = 0
        next = IntArray(n + 1)

        println(solution(n))

        val sb = StringBuilder("$n ")
        var nextValue = next[n]
        while (nextValue != 0) {
            sb.append("$nextValue ")
            nextValue = next[nextValue]
        }

        print(sb)
    }

    private fun solution(n: Int): Int {
        if (dp[n] != max) return dp[n]

        if (n % 3 == 0) update(n, n / 3, solution(n / 3))
        if (n % 2 == 0) update(n, n / 2, solution(n / 2))
        update(n, n - 1, solution(n - 1))

        return dp[n]
    }

    private fun update(n: Int, nextN: Int, result: Int) {
        if (result + 1 >= dp[n]) return

        dp[n] = result + 1
        next[n] = nextN
    }
}

fun main() {
    B_12852()()
}
